test/test_type_promotion.py

```diff
@@ -430,6 +431,104 @@ def test_transpose(self, device):
+        (dense2, sparse2) = self._test_sparse_op_input_tensors(device, dtype2, coalesced, op_name != 'div')
+        common_dtype = torch.result_type(dense1, dense2)
+        if self.device_type == 'cpu' and common_dtype == torch.half:
+            self.assertRaises(RuntimeError, lambda: op(s1, d2))
+
+        if inplace and not torch.can_cast(common_dtype, dtype1):
+            self.assertRaises(RuntimeError, lambda: op(dense1, sparse2))
+            self.assertRaises(RuntimeError, lambda: op(sparse1, sparse2))
+            self.assertRaises(RuntimeError, lambda: op(sparse1, dense2))
+            return
+
+        expected = op(dense1.clone(), dense2)
+        precision = self._get_precision(expected.dtype, coalesced)
+        test_tensors = [expected, dense1, sparse1, dense2, sparse2]
+        e, d1, s1, d2, s2 = [x.clone() for x in test_tensors] if inplace else test_tensors
+
+        # Test op(sparse, sparse)
+        if op_name != 'div':
+            sparse = op(s1, s2)
+            self.assertEqual(sparse.dtype, e.dtype)
+            self.assertEqual(e, sparse.to_dense(), prec=precision, message=err)
+        else:
+            # sparse division only supports division by a scalar
+            self.assertRaises(RuntimeError, lambda: op(s1, s2).to_dense())
+
+        # Test op(dense, sparse)
+        if add_sub:
+            if inplace:
+                e, d1, s1, d2, s2 = [x.clone() for x in test_tensors]
+            dense_sparse = op(d1, s2)
+            self.assertEqual(e, dense_sparse, prec=precision, message=err)
+        else:
+            # sparse division only supports division by a scalar
+            # mul: Didn't find kernel to dispatch to for operator 'aten::_nnz'
+            self.assertRaises(RuntimeError, lambda: op(d1, s2))
+
+        # Test op(sparse, dense) not supported for any ops:
+        # add(sparse, dense) is not supported. Use add(dense, sparse) instead.
+        # sparse division only supports division by a scalar
+        # mul: Didn't find kernel to dispatch to for operator 'aten::_nnz'.
+        self.assertRaises(RuntimeError, lambda: op(s1, d2))
+
+        # Test op(sparse, scalar)
+        if not add_sub and not (self.device_type == 'cpu' and dtype1 == torch.half):
+            if inplace:
+                e, d1, s1, d2, s2 = [x.clone() for x in test_tensors]
+            scalar = d2.view(d2.numel())[0].item()
+
+            sparse = op(s1, scalar)
+            dense_scalar = op(d1, scalar)
+            self.assertEqual(sparse.dtype, dense_scalar.dtype)
+            self.assertEqual(dense_scalar, sparse.to_dense(), prec=precision, message=err)
+        else:
+            # add(sparse, dense) is not supported. Use add(dense, sparse) instead.
+            # "mul_cpu" / "div_cpu" not implemented for 'Half'
+            self.assertRaises(RuntimeError, lambda: op(s1, d2.view(d2.numel())[0].item()))
+
+    @onlyOnCPUAndCUDA
+    def test_sparse_ops(self, device):
```

Comment:
How long does this test take?

Reply:
almost a minute

Reply:
It's generally bad for a single unit test to take one minute. I know this single test is a conglomeration of some number of tests though (108? 400? not sure) but the longer the test is, the greater the risk for (1) flakiness (2) being deemed a "slow test" and being delegated to not run on PRs.

Reply:
Split this into four calling functions, one each for add/sub/mul/div so that no one test runs so long.

Reply:
Are we doing more testing than necessary? Is there any way to cut down on the total test time?

Let's assume that we have an op `op`. There are maybe `N` tests for op, where N is the number of dtypes. If we implement type promotion on `op`, then we end up testing `N x N` cases. This means that in the future, if all of our operations support type promotion, then we will end up with a quadratic number of tests that we have today. This doesn't sound scalable to me.

Reply:
Do you have insight into how numpy does their testing for this?

Reply:
One thing I am thinking is that we could test all "dtype1-dtype2" combinations where "dtype1 = dtype2", test one type promotion combination (like float - double?) and run the rest in the slow ci.

Reply:
It's a lot, but I think this actually is significant meaningful coverage, and I'm planning to make another pass later where I delete a number of tests that overlap with the added coverage here. We've missed a lot of issues (two that were caught by this test: atomic add and uncoalesced div) that only affected a single dtype, or only when uncoalesced, etc. 

We don't necessarily need to add every promoting op -- I'd think we want better coverage for ops as basic as arithmetic.

There might also not be so many binary ops, but if we find this gets too slow we could turn some of this off later too.

Reply:
I'm used to unit tests taking less than 1 second to run.

Because of that, I'm having a hard time evaluating whether or not adding 60s to testing is cause for worry or not (with each test taking 15s). How long does a typical run of test/test_torch.py take to run, in comparison?

Reply:
the full `python test_torch.py` takes quite a while to run, not sure exactly but certainly more than a few minutes.

I can certainly understand the concern of adding a lot of time to the test runs. I weigh that partly against the amount of coverage gained though, and to me this seems like pretty substantial meaningful coverage.

Reply:
Do we need inputs of size 10x10x10 (1000 values) to run for all of the tests though? Could we get away with something like 5 x 5 x 5 (125 values)? I'm not actually sure how many random tensors we are generating

Reply:
I agree 5x5x5 is probably also plenty.